Require JDK 11, use `javac --release` to validate Java API (Java 11+, Gradle 4.4+)

msgilligan opened the pull request:
See Issue #2149. This PR updates the minimum JDK to 11, but the build **still works with Debian Gradle** 4.4.1.

By taking advantages of features in JDK 11+ we get better validation and (in a future commit using Gradle 6.7+) test coverage for the older JDKs.

* Require JDK 11 for the build
* Use `javac -release` option to validate API for older JDK when compiling. (e.g. for —release 8, no JDK 9+ APIs will be allowed)
* Simplify settings.gradle now that we require JDK 11
* Remove JDK 8 from Github Actions matrix

With newer Gradle versions the build is progressively enhanced:

* Gradle 4.4+ Minimum Gradle required
* Gradle 4.10+ wallettemplate is built (JavaFX)

UPDATE: The following are removed from this PR and will be added in a subsequent PR:

* If Gradle 6.7+, use javaToolchains to test on older JDKs (Tests will be run on Gradle’s JDK and also JDK 7, JDK 8)

msgilligan commented:
This PR will need a squash and rebase.  (It currently duplicates changes that are in PR #2146, #2147)

msgilligan commented:
@schildbach Rebased (which dramatically simplifies the diff) and the lead comment has been rewritten to be up-to-date and more clear.

schildbach commented:
If we merge https://github.com/bitcoinj/bitcoinj/pull/2221, I guess then the requirement to test for Java 7 API compatibility would be gone. However we still need a test for Android API compatibility.

msgilligan commented:
> However we still need a test for Android API compatibility.

I think the best way to do this would be to create a new GitHub Actions Workflow that loads Android tooling (lint?) to do the check.

schildbach commented:
> > However we still need a test for Android API compatibility.
> 
> I think the best way to do this would be to create a new GitHub Actions Workflow that loads Android tooling (lint?) to do the check.

I'll watch out for some instructions to do this, either on Gitlab or here.

msgilligan commented:
> I'll watch out for some instructions to do this, either on Gitlab or here.

We might even be able to run some integration tests in an Android simulator.

schildbach commented:
We can probably re-use this PR for the next cycle, when we require the Java 11 SDK for building and need to check for Java 8 API compatibility.

msgilligan commented on core/build.gradle:
```diff
@@ -25,19 +27,23 @@ dependencies {
 }
 
-test {
+tasks.withType(Test) {
```
No, but it is required for testing on multiple JDKs -- do you want 1 PR for upgrading-to/requiring JDK 11 and another for running the tests on older JDKs or just one PR?

msgilligan replied:
I removed testing with JDK 8, but left the `task.withType(Test)` as it allows us to add the JDK 8 test more easily in the future.

And as we discussed we do want to add it, although we may want to fine-tune how it is activated.

msgilligan commented on core/build.gradle:
```diff
@@ -47,6 +53,30 @@ test {
+if (GradleVersion.current().compareTo(gradleVersionToolchains) > 0) {
+    // If the Gradle Java Toolchains feature is available, run tests on older JDKs
+    System.err.println "Enabling tests on older JDKs, because ${GradleVersion.current()}"
+    task('testOnJdk7', type: Test) {
```
I will remove this.

msgilligan commented on settings.gradle:
```diff
@@ -21,10 +25,8 @@ project(':tools').name = 'bitcoinj-tools'
 
-if (GradleVersion.current().compareTo(minFxGradleVersion) >= 0 && JavaVersion.current().isJava11Compatible()) {
-    System.err.println "Including wallettemplate because ${GradleVersion.current()} and Java ${JavaVersion.current()}"
+if (GradleVersion.current().compareTo(minFxGradleVersion) > 0) {
```
OK

msgilligan commented:
I force pushed a simple rebase which failed to build on recent Gradle/Protobuf: https://github.com/bitcoinj/bitcoinj/commit/544c0f1db73289d0977949f65bb7a52362788c97

I then pushed a commit that fully upgrades core to JDK 8 and addresses your comments and everything is building: https://github.com/bitcoinj/bitcoinj/commit/544c0f1db73289d0977949f65bb7a52362788c97

I can squash them if you like.

Otherwise, how does this look?

schildbach commented on README.md:
```diff
@@ -22,13 +22,13 @@ To get started, it is best to have the latest JDK and Gradle installed. The HEAD
 #### Building from the command line
 
-Official builds are currently using with JDK 8, even though the `core` module is compatible with JDK 7 and later.
+Official builds are currently using JDK 11, even though the `core` module is compatible with JDK 8 and later.
```
I wonder if we should perhaps remove this line? It confuses me because I think this is the build requirement, but actually it seems to describe the API requirement, which is already described in the paragraph above.

msgilligan replied:
I agree. I'll remove it and squash when I'm next at my computer.

schildbach commented:
Feel free to squash the commits, but I'll can do it too.

msgilligan commented:
"Gradle javaToolchain" and "test on JDK8" have been removed from the PR and from its title and will be added in a separate PR

msgilligan commented:
OK, @schildbach. Updated, squashed, and ready to merge!

schildbach commented:
I merged this, but I removed the `task.withType(Test)` change too. I think the change is not related to the goal of requiring JDK 11.

Imho the change can go into the PR that actually requires it.

msgilligan commented:
> I merged this, but I removed the `task.withType(Test)` change too. I think the change is not related to the goal of requiring JDK 11.
> 
> 
> 
> Imho the change can go into the PR that actually requires it.

Fair enough.
